src/main/java/edu/cmu/oli/content/svnmanager/SVNManager.java

```diff
@@ -342,8 +343,12 @@ public boolean isVersioned(File path) {
         return fileList;
     }
 
-    public List<File> listAddedFiles(File path) throws SVNException {
-        SVNClientManager svnClientManager = SVNClientManager.newInstance();
+    public synchronized List<File> listAddedFiles(File path) throws SVNException {
+        SVNClientManager svnClientManager = clientManagerMap.get(path.getAbsolutePath());
+        if(svnClientManager == null) {
```

Comment:
The use of `ConcurrentHashMap` here alone doesn't make this implementation "correct" with regards to thread safety.  With a context switch at just the right time, two different threads will see the `svnClientManager` as `null` and then create and put a new instance. A thread safe solution here is something like: 

```
SVNClientManager svnClientManager;
synchronized (clientManagerMap) {
  svnClientManager = clientManagerMap.get(path.getAbsolutePath());
  if (svnClientManager == null) {
    svnClientManager = SVNClientManager.newInstance();
    clientManagerMap.put(path.getAbsolutePath(), svnClientManager);
  }
}
```

Reply:
Ah. Makes sense. Updated with suggested changes.